### PR TITLE
Admin Router: Refromat and update documentation

### PR DESCRIPTION
**Pull request by vespian**
## High-Level Description

This PR:
* Re-adjusts some of the paragraphs in README.md on the 80 cols limit.
* Updates the documentation with the latest features&changes to the test harness

Semantic line feeds are something we can discuss at some point. My main problem is that I very often work with documentation using Vim. 80cols limit makes it easier and more readable.

Tests are expected to fail.

## Related issues:
https://jira.mesosphere.com/browse/DCOS_OSS-1012 Admin Router: Quick-start guide for writing tests

## Related PRs:
AR-NEXT: https://github.com/dcos/dcos/pull/1802

### CODE REVIEW

File: packages/adminrouter/extra/src/README.md [hunk 1]
```diff
@@ -134,14 +137,14 @@ includes
...
 
-All NGINX related configuration directives reside in the `includes` directory which
-contains directories reflecting all the sections present in NGINX configuration
-(http|main|server). Because they are always present, no
-matter the flavour/server type, they are chosen as top-level directories
-in the `includes` dir. The `common.conf`, `agent.conf` and `master.conf`
-files are flavour-agnostic thus they reside in each section's main dir
-and are present/included only in the Open AR repository. EE repository re-uses them
-after being applied on top of the Open repository. Contents of each of the
+All the NGINX related configuration directives reside in the `includes`
```
Comment (vespian) [inline packages/adminrouter/extra/src/README.md:140]: Just the reformat.

File: packages/adminrouter/extra/src/README.md [hunk 2]
```diff
@@ -214,56 +217,58 @@ http {
 ```
 
 ### Enforcing code reuse
-The EE repository contains only EE directories, all common/agent-common/master-common
-code resides in Open repository. This way EE repository becomes an overlay on
-top of Open. Only `nginx.(master|agent).conf` are overwritten while applying EE
-repository on top of Open during DC/OS image build. EE DC/OS image build scripts
-remove all open directories from the Open repository before applying EE repository on
-top of it.
+The EE repository contains only EE directories, all
```
Comment (vespian) [inline packages/adminrouter/extra/src/README.md:220]: Just the reformat.

File: packages/adminrouter/extra/src/README.md [hunk 3]
```diff
@@ -214,56 +217,58 @@ http {
...
 duplication (developers can simply start putting copies of code to both
 `open/` and `ee/` directories) but it makes it easier to reuse the
 code and encourages good behaviours.
 
 ### Lua code deduplication
-It is not strictly required to provide the same level of flexibility for Lua code
-as it is the case for NGINX includes and thus it's possible to simplify the code
-a bit. It is sufficient to just differentiate Lua code basing on the repository flavour.
-Both agents and masters can share the same Lua code.
+It is not required to provide the same level of flexibility for Lua code as it
```
Comment (vespian) [inline packages/adminrouter/extra/src/README.md:233]: Just the reformat.

File: packages/adminrouter/extra/src/README.md [hunk 4]
```diff
@@ -214,56 +217,58 @@ http {
...
-   `ee.lua|open.lua|common.lua` parts which is described in next bullet point.
-   Depending on the flavour, either `ee.lua` or `open.lua` is imported and correct
-   argument-less function is used. This approach also enables to share some of
-   the NGINX `location` blocks between flavours - `location` code is the same,
-   even though the Lua code used by the block differs.
- * some libraries/functions are structured in a way that extracts parts common for
-   both flavours and `ee/open` parts that are included only in EE and Open repos
-   respectively. An example of this approach is auth library which is splitted
-   into three parts:
+   functions for the NGINX configuration. They translate the original call into
```
Comment (vespian) [inline packages/adminrouter/extra/src/README.md:249]: Just the reformat.

File: packages/adminrouter/extra/src/README.md [hunk 5]
```diff
@@ -214,56 +217,58 @@ http {
...
       shared by both EE and Open code.
     * `lib/auth/ee.lua` - present only in EE repository, with functions specific
       to EE that use boilerplate from lib/auth/common.lua.
     * `lib/auth/open.lua` - as above but for Open repository.
-   `init_by_lua` OpenResty call in `includes/http/(open|ee)/common.conf` imports
-   `auth.open` or `auth.ee` modules as auth respectively.  The module is registered in
-   the global namespace so all other Lua code uses it. This approach also allows
-   for some degree of code separation enforcement as Open lua libs are removed
-   during EE repository apply.
+   `init_by_lua` OpenResty call in `includes/http/(open|ee)/common.conf`
```
Comment (vespian) [inline packages/adminrouter/extra/src/README.md:266]: Just the reformat.

File: packages/adminrouter/extra/src/README.md [hunk 6]
```diff
@@ -450,10 +456,10 @@ the host running the task using the one of the ports allocated to the task.
 The chosen port is defined by the `DCOS_SERVICE_PORT_INDEX` label.
 
 In order for the forwarding to work reliably across task failures, we recommend
-co-locating the endpoints with the task. This way, if the task is restarted on a
-potentially different host and with different ports, Admin Router will pick up the
-new labels and update the routing. Due to caching there might be an up to
-30-second delay before the new routing is working.
+co-locating the endpoints with the task. This way, if the task is restarted on
```
Comment (vespian) [inline packages/adminrouter/extra/src/README.md:459]: Just the reformat.

File: packages/adminrouter/extra/src/README.md [hunk 7]
```diff
@@ -578,7 +584,8 @@ shared memory will still be considered fresh.
 The `freshness` of the cache is governed by few variables:
 * `CACHE_EXPIRATION` - if the age of the cached data is smaller than
   `CACHE_EXPIRATION` seconds, then the cache refresh will not occur if it's
-  ngx.timer context and request processing code will use the data stored in shared memory.
+  ngx.timer context and request processing code will use the data stored in
```
Comment (vespian) [inline packages/adminrouter/extra/src/README.md:587]: Just the reformat.

File: packages/adminrouter/extra/src/README.md [hunk 8]
```diff
@@ -595,10 +602,10 @@ The relation between these is:
 
 The reason why we put `<<` in front of `CACHE_MAX_AGE_HARD_LIMIT` is to make
 the cache a bit of a "best-effort" one - In the case when Mesos and/or Marathon
-dies, the cache should still be able to serve data for a reasonable amount of time
-and thus give the operator some time to solve the underlying issue. For example
-Mesos tasks do not move that often and the data stored in NGINX should still be
-usable, at least partially.
+dies, the cache should still be able to serve data for a reasonable amount of
```
Comment (vespian) [inline packages/adminrouter/extra/src/README.md:605]: Just the reformat.

File: packages/adminrouter/extra/src/README.md [hunk 9]
```diff
@@ -608,16 +615,17 @@ well. In order to coordinate it, locking was introduced.
 
 There are two different locking behaviours, depending on the context from which
 the update was triggered:
-* for timer-based refreshes, the lock is non-blocking. If there is an update already
-  in progress, execution is aborted, and next timer-based refresh is scheduled.
-* in case of request-triggered update, lock is blocking and the lock timeout
-  is equal to `CACHE_REFRESH_LOCK_TIMEOUT` seconds. This way, during
-  `<0, CACHE_FIRST_POLL_DELAY>` period, requests are queued while waiting for the
+* for timer-based refreshes, the lock is non-blocking. If there is an update
```
Comment (vespian) [inline packages/adminrouter/extra/src/README.md:618]: Just the reformat.

File: packages/adminrouter/extra/src/README.md [hunk 10]
```diff
@@ -659,46 +666,45 @@ the dependencies.
 
 ### Makefile
 Makefile provides an easy way to start the testing environment without the need
-to worry about the correct docker commands. Its core concept is `adminrouter-devkit`
-container which is contains all the dependencies that are needed to run
-Admin Router, and inside which all tests-related commands are run.
+to worry about the correct docker commands. Its core concept is
```
Comment (vespian) [inline packages/adminrouter/extra/src/README.md:669]: Just the reformat.

File: packages/adminrouter/extra/src/README.md [hunk 11]
```diff
@@ -659,46 +666,45 @@ the dependencies.
...
 It exposes a couple of targets:
 * `make clean` - remove all containers created by the test harness. It does not
   remove the images themselves though as the layer cache may be useful later
   on and the user may remove them themselves.
 * `make devkit` - creates the `adminrouter-devkit` container. By default other
-   targets execute this step automatically if devkit container image does not
-   exist yet.
-* `make update-devkit` - updates `adminrouter-devkit`. Should be run every time
-   the Dockerfile or its dependencies change.
+  targets execute this step automatically if the devkit container image does not
```
Comment (vespian) [inline packages/adminrouter/extra/src/README.md:679]: Just the reformat.

File: packages/adminrouter/extra/src/README.md [hunk 12]
```diff
@@ -659,46 +666,45 @@ the dependencies.
...
-container. It follows the same build process for NGINX that happens during
-DC/OS build with the exception of setting the  `--with-debug` flag. It also
-contains some basic debugging tools, pytest related dependencies and
-files that help pytest mimic the DC/OS environment. Their location is then
-exported via environment variables to the pytest code, so changing their location
-can be done by edition only the Dockerfile.
-
-Due to some issues with overlayfs on older kernels, the container links
-/tmp directory to /dev/shm.
+As mentioned earlier, all the commands are executed inside the
```
Comment (vespian) [inline packages/adminrouter/extra/src/README.md:693]: Just the reformat.

File: packages/adminrouter/extra/src/README.md [hunk 13]
```diff
@@ -659,46 +666,45 @@ the dependencies.
...
+so changing their location can be done by edition only the Dockerfile.
 
 ### Repository flavours
-Admin Router repository can come in two variants/flavours - Enterprise and Opensource.
-Tests determine it basing on the directory structure - if tests contain
-`test-harness/tests/open/` directory then the repo is treated as Opensource one,
-and in case when `test-harness/tests/ee/` directory is present - as the
-enterprise one. Pytest fixture `repo_is_ee` takes care of it and is pulled in
-as dependencies by all the fixtures that need this information.
+Admin Router repository can come in two variants/flavours - Enterprise and
```
Comment (vespian) [inline packages/adminrouter/extra/src/README.md:702]: Just the reformat.

File: packages/adminrouter/extra/src/README.md [hunk 14]
```diff
@@ -838,29 +845,28 @@ with the line length limit hard-coded to 4096 bytes.
...
-need to have custom lifetime or just single-test scoped lifetime, then it's
-necessary to use `nginx_class` fixture instead of simple `master_ar_process` or
-`agent_ar_process` ones.
+need to have custom lifetime, then it's necessary to use `nginx_class`
+fixture which provides maximum flexibility.
 
 NGINX instances have the `.make_url_from_path` method which is a convenient way
-to generate AR URLs for tests. It uses exhibitor endpoint as it's present in
-all DC/OS configurations and uses auth features as well.
+to generate AR URLs for tests. Users do not have to worry whether this is an
```
Comment (vespian) [inline packages/adminrouter/extra/src/README.md:852]: Just the reformat.

File: packages/adminrouter/extra/src/README.md [hunk 15]
```diff
@@ -8,7 +8,10 @@ within the cluster.
 
 ## Routes
 
-Admin Router runs on both master and agent nodes, each with different configurations. From these NGINX config files, [Ngindox](https://github.com/karlkfi/ngindox) is used to generates swagger-like docs:
+Admin Router runs on both master and agent nodes, each with different
+configurations. From these NGINX config files,
```
Comment (adamtheturtle) [inline packages/adminrouter/extra/src/README.md:12]: `s/different/different NGINX`. This makes it clear which files the next sentence refers to as I had to re-read.

**Review by adamtheturtle — COMMENTED** [review]
I am ignoring existing grammatical and language issues as per our previous discussion. If you would like me to review the existing language, please let me know and I will have another look over.

**Review by gpaul — APPROVED** [review]
LGTM

File: packages/adminrouter/extra/src/README.md [hunk 16]
```diff
@@ -838,29 +845,28 @@ with the line length limit hard-coded to 4096 bytes.
...
+to generate AR URLs for tests. Users do not have to worry whether this is an
+agent or master instance nor know the TCP port that given instance listens on.
 
-#### pytest tests
-Directly from the fact that Admin Router subprocess fixture is module scoped,
-stems some of the current structure of the tests directory. There are three
-different files where tests reside:
+#### AR instance reuse
+Some of the tests can share the same AR instance as it is not being mutated by
+them. There are currently three files which contain:
```
Comment (adamtheturtle) [inline packages/adminrouter/extra/src/README.md:857]: I think that there is a missing sentence segment after "contain".

**Review by adamtheturtle — APPROVED** [review]
Thanks @vespian . As per other comments, there are multiple language issues which I would like to fix but they are not new in this change. Changes to this document are also hard to review because every added word might require a reformat of a whole paragraph. I would like to remove this difficulty at some point in the near future.

File: packages/adminrouter/extra/src/README.md (hunk 16 above)
Comment (vespian) [reply packages/adminrouter/extra/src/README.md:857]:
```
  856 Some of the tests can share the same AR instance as it is not being mutated by
~ 857 them. There are currently at least three files which contain tests which share
+ 858 AR instance:
```

**Comment by vespian** [comment]
> Changes to this document are also hard to review because every added word might require a reformat of a whole paragraph. I would like to remove this difficulty at some point in the near future.

Agreed, but let's do it in a separate PR, track it with an issue, and ask docs team if they are up for the task.

**Pull request merged**